// @angular/compiler-cli/src/ngtsc/resource_loader.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/resource_loader", ["require", "exports", "fs"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var fs = require("fs");
    /**
     * `ResourceLoader` which delegates to a `CompilerHost` resource loading method.
     */
    var HostResourceLoader = /** @class */ (function () {
        function HostResourceLoader(host) {
            this.host = host;
            this.cache = new Map();
            this.fetching = new Set();
        }
        HostResourceLoader.prototype.preload = function (url) {
            var _this = this;
            if (this.cache.has(url) || this.fetching.has(url)) {
                return undefined;
            }
            var result = this.host(url);
            if (typeof result === 'string') {
                this.cache.set(url, result);
                return undefined;
            }
            else {
                this.fetching.add(url);
                return result.then(function (str) {
                    _this.fetching.delete(url);
                    _this.cache.set(url, str);
                });
            }
        };
        HostResourceLoader.prototype.load = function (url) {
            if (this.cache.has(url)) {
                return this.cache.get(url);
            }
            var result = this.host(url);
            if (typeof result !== 'string') {
                throw new Error("HostResourceLoader: host(" + url + ") returned a Promise");
            }
            this.cache.set(url, result);
            return result;
        };
        return HostResourceLoader;
    }());
    exports.HostResourceLoader = HostResourceLoader;
    /**
     * `ResourceLoader` which directly uses the filesystem to resolve resources synchronously.
     */
    var FileResourceLoader = /** @class */ (function () {
        function FileResourceLoader() {
        }
        FileResourceLoader.prototype.load = function (url) { return fs.readFileSync(url, 'utf8'); };
        return FileResourceLoader;
    }());
    exports.FileResourceLoader = FileResourceLoader;
});
