// @angular/core/esm5/src/ivy_switch/runtime/legacy.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as ivyOn from './ivy_switch_on';
function noopFactory() {
    var tokens = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        tokens[_i] = arguments[_i];
    }
}
export var R3_ELEMENT_REF_FACTORY__POST_NGCC__ = ivyOn.R3_ELEMENT_REF_FACTORY;
export var R3_TEMPLATE_REF_FACTORY__POST_NGCC__ = ivyOn.R3_TEMPLATE_REF_FACTORY;
export var R3_CHANGE_DETECTOR_REF_FACTORY__POST_NGCC__ = ivyOn.R3_CHANGE_DETECTOR_REF_FACTORY;
export var R3_VIEW_CONTAINER_REF_FACTORY__POST_NGCC__ = ivyOn.R3_VIEW_CONTAINER_REF_FACTORY;
export var R3_RENDERER2_FACTORY__POST_NGCC__ = ivyOn.R3_RENDERER2_FACTORY;
export var R3_ELEMENT_REF_FACTORY__PRE_NGCC__ = noopFactory;
export var R3_TEMPLATE_REF_FACTORY__PRE_NGCC__ = noopFactory;
export var R3_CHANGE_DETECTOR_REF_FACTORY__PRE_NGCC__ = noopFactory;
export var R3_VIEW_CONTAINER_REF_FACTORY__PRE_NGCC__ = noopFactory;
export var R3_RENDERER2_FACTORY__PRE_NGCC__ = noopFactory;
export var R3_ELEMENT_REF_FACTORY = R3_ELEMENT_REF_FACTORY__PRE_NGCC__;
export var R3_TEMPLATE_REF_FACTORY = R3_TEMPLATE_REF_FACTORY__PRE_NGCC__;
export var R3_CHANGE_DETECTOR_REF_FACTORY = R3_CHANGE_DETECTOR_REF_FACTORY__PRE_NGCC__;
export var R3_VIEW_CONTAINER_REF_FACTORY = R3_VIEW_CONTAINER_REF_FACTORY__PRE_NGCC__;
export var R3_RENDERER2_FACTORY = R3_RENDERER2_FACTORY__PRE_NGCC__;
