// @angular/compiler-cli/src/language_services.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/language_services", ["require", "exports", "@angular/compiler-cli/src/diagnostics/expression_diagnostics", "@angular/compiler-cli/src/diagnostics/expression_type", "@angular/compiler-cli/src/diagnostics/symbols", "@angular/compiler-cli/src/diagnostics/typescript_symbols", "@angular/compiler-cli/src/metadata/index", "@angular/compiler-cli/src/transformers/metadata_reader"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /*
    
    The API from compiler-cli that language-service can see.
    It is important that none the exported modules require anything other than
    Angular modules and Typescript as this will indirectly add a dependency
    to the language service.
    
    */
    var expression_diagnostics_1 = require("@angular/compiler-cli/src/diagnostics/expression_diagnostics");
    exports.getExpressionDiagnostics = expression_diagnostics_1.getExpressionDiagnostics;
    exports.getExpressionScope = expression_diagnostics_1.getExpressionScope;
    exports.getTemplateExpressionDiagnostics = expression_diagnostics_1.getTemplateExpressionDiagnostics;
    var expression_type_1 = require("@angular/compiler-cli/src/diagnostics/expression_type");
    exports.AstType = expression_type_1.AstType;
    exports.DiagnosticKind = expression_type_1.DiagnosticKind;
    exports.TypeDiagnostic = expression_type_1.TypeDiagnostic;
    var symbols_1 = require("@angular/compiler-cli/src/diagnostics/symbols");
    exports.BuiltinType = symbols_1.BuiltinType;
    var typescript_symbols_1 = require("@angular/compiler-cli/src/diagnostics/typescript_symbols");
    exports.getClassFromStaticSymbol = typescript_symbols_1.getClassFromStaticSymbol;
    exports.getClassMembers = typescript_symbols_1.getClassMembers;
    exports.getClassMembersFromDeclaration = typescript_symbols_1.getClassMembersFromDeclaration;
    exports.getPipesTable = typescript_symbols_1.getPipesTable;
    exports.getSymbolQuery = typescript_symbols_1.getSymbolQuery;
    var metadata_1 = require("@angular/compiler-cli/src/metadata/index");
    exports.MetadataCollector = metadata_1.MetadataCollector;
    var metadata_reader_1 = require("@angular/compiler-cli/src/transformers/metadata_reader");
    exports.createMetadataReaderCache = metadata_reader_1.createMetadataReaderCache;
    exports.readMetadata = metadata_reader_1.readMetadata;
});
