// @angular/compiler-cli/src/ngtsc/factories/src/host.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/factories/src/host", ["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * A wrapper around a `ts.CompilerHost` which supports generated files.
     */
    var GeneratedFactoryHostWrapper = /** @class */ (function () {
        function GeneratedFactoryHostWrapper(delegate, generator, factoryToSourceMap) {
            this.delegate = delegate;
            this.generator = generator;
            this.factoryToSourceMap = factoryToSourceMap;
            if (delegate.resolveTypeReferenceDirectives) {
                this.resolveTypeReferenceDirectives = function (names, containingFile) {
                    return delegate.resolveTypeReferenceDirectives(names, containingFile);
                };
            }
        }
        GeneratedFactoryHostWrapper.prototype.getSourceFile = function (fileName, languageVersion, onError, shouldCreateNewSourceFile) {
            var canonical = this.getCanonicalFileName(fileName);
            if (this.factoryToSourceMap.has(canonical)) {
                var sourceFileName = this.getCanonicalFileName(this.factoryToSourceMap.get(canonical));
                var sourceFile = this.delegate.getSourceFile(sourceFileName, languageVersion, onError, shouldCreateNewSourceFile);
                if (sourceFile === undefined) {
                    return undefined;
                }
                return this.generator.factoryFor(sourceFile, fileName);
            }
            return this.delegate.getSourceFile(fileName, languageVersion, onError, shouldCreateNewSourceFile);
        };
        GeneratedFactoryHostWrapper.prototype.getDefaultLibFileName = function (options) {
            return this.delegate.getDefaultLibFileName(options);
        };
        GeneratedFactoryHostWrapper.prototype.writeFile = function (fileName, data, writeByteOrderMark, onError, sourceFiles) {
            return this.delegate.writeFile(fileName, data, writeByteOrderMark, onError, sourceFiles);
        };
        GeneratedFactoryHostWrapper.prototype.getCurrentDirectory = function () { return this.delegate.getCurrentDirectory(); };
        GeneratedFactoryHostWrapper.prototype.getDirectories = function (path) { return this.delegate.getDirectories(path); };
        GeneratedFactoryHostWrapper.prototype.getCanonicalFileName = function (fileName) {
            return this.delegate.getCanonicalFileName(fileName);
        };
        GeneratedFactoryHostWrapper.prototype.useCaseSensitiveFileNames = function () { return this.delegate.useCaseSensitiveFileNames(); };
        GeneratedFactoryHostWrapper.prototype.getNewLine = function () { return this.delegate.getNewLine(); };
        GeneratedFactoryHostWrapper.prototype.fileExists = function (fileName) {
            return this.factoryToSourceMap.has(fileName) || this.delegate.fileExists(fileName);
        };
        GeneratedFactoryHostWrapper.prototype.readFile = function (fileName) { return this.delegate.readFile(fileName); };
        return GeneratedFactoryHostWrapper;
    }());
    exports.GeneratedFactoryHostWrapper = GeneratedFactoryHostWrapper;
});
