// codelyzer/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var angularWhitespaceRule_1 = require("./angularWhitespaceRule");
exports.AngularWhitespaceRule = angularWhitespaceRule_1.Rule;
var bananaInBoxRule_1 = require("./bananaInBoxRule");
exports.BananaInBoxRule = bananaInBoxRule_1.Rule;
var componentClassSuffixRule_1 = require("./componentClassSuffixRule");
exports.ComponentClassSuffixRule = componentClassSuffixRule_1.Rule;
var componentSelectorRule_1 = require("./componentSelectorRule");
exports.ComponentSelectorRule = componentSelectorRule_1.Rule;
var contextualLifeCycleRule_1 = require("./contextualLifeCycleRule");
exports.ContextualLifeCycleRule = contextualLifeCycleRule_1.Rule;
var decoratorNotAllowedRule_1 = require("./decoratorNotAllowedRule");
exports.DecoratorNotAllowedRule = decoratorNotAllowedRule_1.Rule;
var directiveClassSuffixRule_1 = require("./directiveClassSuffixRule");
exports.DirectiveClassSuffixRule = directiveClassSuffixRule_1.Rule;
var directiveSelectorRule_1 = require("./directiveSelectorRule");
exports.DirectiveSelectorRule = directiveSelectorRule_1.Rule;
var enforceComponentSelectorRule_1 = require("./enforceComponentSelectorRule");
exports.EnforceComponentSelectorRule = enforceComponentSelectorRule_1.Rule;
var i18nRule_1 = require("./i18nRule");
exports.I18nRule = i18nRule_1.Rule;
var importDestructuringSpacingRule_1 = require("./importDestructuringSpacingRule");
exports.ImportDestructuringSpacingRule = importDestructuringSpacingRule_1.Rule;
var maxInlineDeclarationsRule_1 = require("./maxInlineDeclarationsRule");
exports.MaxInlineDeclarationsRule = maxInlineDeclarationsRule_1.Rule;
var noAttributeParameterDecoratorRule_1 = require("./noAttributeParameterDecoratorRule");
exports.NoAttributeParameterDecoratorRule = noAttributeParameterDecoratorRule_1.Rule;
var noConflictingLifeCycleHooksRule_1 = require("./noConflictingLifeCycleHooksRule");
exports.NoConflictingLifeCycleHooksRule = noConflictingLifeCycleHooksRule_1.Rule;
var noForwardRefRule_1 = require("./noForwardRefRule");
exports.NoForwardRefRule = noForwardRefRule_1.Rule;
var noInputPrefixRule_1 = require("./noInputPrefixRule");
exports.NoInputPrefixRule = noInputPrefixRule_1.Rule;
var noInputRenameRule_1 = require("./noInputRenameRule");
exports.NoInputRenameRule = noInputRenameRule_1.Rule;
var noLifeCycleCallRule_1 = require("./noLifeCycleCallRule");
exports.NoLifeCycleCallRule = noLifeCycleCallRule_1.Rule;
var noOutputNamedAfterStandardEventRule_1 = require("./noOutputNamedAfterStandardEventRule");
exports.NoOutputNamedAfterStandardEventRule = noOutputNamedAfterStandardEventRule_1.Rule;
var noOutputOnPrefixRule_1 = require("./noOutputOnPrefixRule");
exports.NoOutputOnPrefixRule = noOutputOnPrefixRule_1.Rule;
var noOutputRenameRule_1 = require("./noOutputRenameRule");
exports.NoOutputRenameRule = noOutputRenameRule_1.Rule;
var noQueriesParameterRule_1 = require("./noQueriesParameterRule");
exports.NoQueriesParameterRule = noQueriesParameterRule_1.Rule;
var noTemplateCallExpressionRule_1 = require("./noTemplateCallExpressionRule");
exports.NoTemplateCallExpressionRule = noTemplateCallExpressionRule_1.Rule;
var noUnusedCssRule_1 = require("./noUnusedCssRule");
exports.NoUnusedCssRule = noUnusedCssRule_1.Rule;
var pipeImpureRule_1 = require("./pipeImpureRule");
exports.PipeImpureRule = pipeImpureRule_1.Rule;
var pipeNamingRule_1 = require("./pipeNamingRule");
exports.PipeNamingRule = pipeNamingRule_1.Rule;
var pipePrefixRule_1 = require("./pipePrefixRule");
exports.PipePrefixRule = pipePrefixRule_1.Rule;
var preferInlineDecoratorRule_1 = require("./preferInlineDecoratorRule");
exports.PreferInlineDecorator = preferInlineDecoratorRule_1.Rule;
var preferOutputReadonlyRule_1 = require("./preferOutputReadonlyRule");
exports.PreferOutputReadonlyRule = preferOutputReadonlyRule_1.Rule;
var templateConditionalComplexityRule_1 = require("./templateConditionalComplexityRule");
exports.TemplateConditionalComplexityRule = templateConditionalComplexityRule_1.Rule;
var templateCyclomaticComplexityRule_1 = require("./templateCyclomaticComplexityRule");
exports.TemplateCyclomaticComplexityRule = templateCyclomaticComplexityRule_1.Rule;
var templatesNoNegatedAsyncRule_1 = require("./templatesNoNegatedAsyncRule");
exports.TemplatesNoNegatedAsync = templatesNoNegatedAsyncRule_1.Rule;
var trackByFunctionRule_1 = require("./trackByFunctionRule");
exports.TrackByFunctionRule = trackByFunctionRule_1.Rule;
var useHostPropertyDecoratorRule_1 = require("./useHostPropertyDecoratorRule");
exports.UseHostPropertyDecoratorRule = useHostPropertyDecoratorRule_1.Rule;
var useInputPropertyDecoratorRule_1 = require("./useInputPropertyDecoratorRule");
exports.UseInputPropertyDecoratorRule = useInputPropertyDecoratorRule_1.Rule;
var useLifeCycleInterfaceRule_1 = require("./useLifeCycleInterfaceRule");
exports.UseLifeCycleInterfaceRule = useLifeCycleInterfaceRule_1.Rule;
var useOutputPropertyDecoratorRule_1 = require("./useOutputPropertyDecoratorRule");
exports.UseOutputPropertyDecoratorRule = useOutputPropertyDecoratorRule_1.Rule;
var usePipeDecoratorRule_1 = require("./usePipeDecoratorRule");
exports.UsePipeDecoratorRule = usePipeDecoratorRule_1.Rule;
var usePipeTransformInterfaceRule_1 = require("./usePipeTransformInterfaceRule");
exports.UsePipeTransformInterfaceRule = usePipeTransformInterfaceRule_1.Rule;
var useViewEncapsulationRule_1 = require("./useViewEncapsulationRule");
exports.UseViewEncapsulationRule = useViewEncapsulationRule_1.Rule;
__export(require("./angular"));
exports.rulesDirectory = '.';
