// gts/build/src/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Copyright 2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const cp = require("child_process");
const fs = require("fs");
const inquirer = require("inquirer");
const path = require("path");
const ncp_1 = require("ncp");
const util = require("util");
const util_1 = require("./util");
const chalk_1 = require("chalk");
const pkg = require('../../package.json');
const ncpp = util.promisify(ncp_1.ncp);
const DEFAULT_PACKAGE_JSON = {
    name: '',
    version: '0.0.0',
    description: '',
    main: 'build/src/index.js',
    types: 'build/src/index.d.ts',
    files: ['build/src'],
    license: 'Apache-2.0',
    keywords: [],
    scripts: { test: 'echo "Error: no test specified" && exit 1' },
};
async function query(message, question, defaultVal, options) {
    if (options.yes) {
        return true;
    }
    else if (options.no) {
        return false;
    }
    if (message) {
        options.logger.log(message);
    }
    const answers = await inquirer.prompt({
        type: 'confirm',
        name: 'query',
        message: question,
        default: defaultVal,
    });
    return answers.query;
}
async function addScripts(packageJson, options) {
    let edits = false;
    const pkgManager = util_1.getPkgManagerCommand(options.yarn);
    const scripts = {
        check: `gts check`,
        clean: 'gts clean',
        compile: `tsc -p .`,
        fix: `gts fix`,
        prepare: `${pkgManager} run compile`,
        pretest: `${pkgManager} run compile`,
        posttest: `${pkgManager} run check`,
    };
    if (!packageJson.scripts) {
        packageJson.scripts = {};
    }
    for (const script of Object.keys(scripts)) {
        let install = true;
        const existing = packageJson.scripts[script];
        const target = scripts[script];
        if (existing !== target) {
            if (existing) {
                const message = `package.json already has a script for ${chalk_1.default.bold(script)}:\n` +
                    `-${chalk_1.default.red(existing)}\n+${chalk_1.default.green(target)}`;
                install = await query(message, 'Replace', false, options);
            }
            if (install) {
                packageJson.scripts[script] = scripts[script];
                edits = true;
            }
        }
    }
    return edits;
}
exports.addScripts = addScripts;
async function addDependencies(packageJson, options) {
    let edits = false;
    const deps = {
        gts: `^${pkg.version}`,
        typescript: pkg.devDependencies.typescript,
        '@types/node': pkg.devDependencies['@types/node'],
    };
    if (!packageJson.devDependencies) {
        packageJson.devDependencies = {};
    }
    for (const dep of Object.keys(deps)) {
        let install = true;
        const existing = packageJson.devDependencies[dep];
        const target = deps[dep];
        if (existing !== target) {
            if (existing) {
                const message = `Already have devDependency for ${chalk_1.default.bold(dep)}:\n` +
                    `-${chalk_1.default.red(existing)}\n+${chalk_1.default.green(target)}`;
                install = await query(message, 'Overwrite', false, options);
            }
            if (install) {
                packageJson.devDependencies[dep] = deps[dep];
                edits = true;
            }
        }
    }
    return edits;
}
exports.addDependencies = addDependencies;
function formatJson(object) {
    // TODO: preserve the indent from the input file.
    const json = JSON.stringify(object, null, '  ');
    return `${json}\n`;
}
async function writePackageJson(packageJson, options) {
    options.logger.log('Writing package.json...');
    if (!options.dryRun) {
        await util_1.writeFileAtomicp('./package.json', formatJson(packageJson));
    }
    const preview = {
        scripts: packageJson.scripts,
        devDependencies: packageJson.devDependencies,
    };
    options.logger.dir(preview);
}
exports.TSLINT_CONFIG = {
    extends: 'gts/tslint.json',
    linterOptions: {
        exclude: ['**/*.json'],
    },
};
async function generateTsLintConfig(options) {
    return generateConfigFile(options, './tslint.json', formatJson(exports.TSLINT_CONFIG));
}
async function generateTsConfig(options) {
    const config = formatJson({
        extends: './node_modules/gts/tsconfig-google.json',
        compilerOptions: { rootDir: '.', outDir: 'build' },
        include: ['src/**/*.ts', 'test/**/*.ts'],
    });
    return generateConfigFile(options, './tsconfig.json', config);
}
async function generatePrettierConfig(options) {
    const style = await util_1.readFilep(path.join(__dirname, '../../prettier.config.js'), 'utf8');
    return generateConfigFile(options, './prettier.config.js', style);
}
async function generateConfigFile(options, filename, contents) {
    let existing;
    try {
        existing = await util_1.readFilep(filename, 'utf8');
    }
    catch (err) {
        if (err.code === 'ENOENT') {
            /* not found, create it. */
        }
        else {
            throw new Error(`Unknown error reading ${filename}: ${err.message}`);
        }
    }
    let writeFile = true;
    if (existing && existing === contents) {
        options.logger.log(`No edits needed in ${filename}`);
        return;
    }
    else if (existing) {
        writeFile = await query(`${chalk_1.default.bold(filename)} already exists`, 'Overwrite', false, options);
    }
    if (writeFile) {
        options.logger.log(`Writing ${filename}...`);
        if (!options.dryRun) {
            await util_1.writeFileAtomicp(filename, contents);
        }
        options.logger.log(contents);
    }
}
async function installDefaultTemplate(options) {
    const cwd = process.cwd();
    const sourceDirName = path.join(__dirname, '../template');
    const targetDirName = path.join(cwd, 'src');
    try {
        fs.mkdirSync(targetDirName);
    }
    catch (error) {
        if (error.code !== 'EEXIST') {
            throw error;
        }
        // Else, continue and populate files into the existing directory.
    }
    // Only install the template if no ts files exist in target directory.
    const files = fs.readdirSync(targetDirName);
    const tsFiles = files.filter(file => file.toLowerCase().endsWith('.ts'));
    if (tsFiles.length !== 0) {
        options.logger.log('Target src directory already has ts files. ' +
            'Template files not installed.');
        return false;
    }
    await ncpp(sourceDirName, targetDirName);
    options.logger.log('Default template installed.');
    return true;
}
exports.installDefaultTemplate = installDefaultTemplate;
async function init(options) {
    let generatedPackageJson = false;
    let packageJson;
    try {
        packageJson = await util_1.readJsonp('./package.json');
    }
    catch (err) {
        if (err.code !== 'ENOENT') {
            throw new Error(`Unable to open package.json file: ${err.message}`);
        }
        const generate = await query(`${chalk_1.default.bold('package.json')} does not exist.`, `Generate`, true, options);
        if (!generate) {
            options.logger.log('Please run from a directory with your package.json.');
            return false;
        }
        packageJson = DEFAULT_PACKAGE_JSON;
        generatedPackageJson = true;
    }
    const addedDeps = await addDependencies(packageJson, options);
    const addedScripts = await addScripts(packageJson, options);
    if (generatedPackageJson || addedDeps || addedScripts) {
        await writePackageJson(packageJson, options);
    }
    else {
        options.logger.log('No edits needed in package.json.');
    }
    await generateTsConfig(options);
    await generateTsLintConfig(options);
    await generatePrettierConfig(options);
    await installDefaultTemplate(options);
    // Run `npm install` after initial setup so `npm run check` works right away.
    if (!options.dryRun) {
        // --ignore-scripts so that compilation doesn't happen because there's no
        // source files yet.
        cp.spawnSync(util_1.getPkgManagerCommand(options.yarn), ['install', '--ignore-scripts'], { stdio: 'inherit' });
    }
    return true;
}
exports.init = init;
